README.rst
==============
 Invenio-Mail
==============

.. image:: https://img.shields.io/travis/inveniosoftware/invenio-mail.svg
        :target: https://travis-ci.org/inveniosoftware/invenio-mail

.. image:: https://img.shields.io/coveralls/inveniosoftware/invenio-mail.svg
        :target: https://coveralls.io/r/inveniosoftware/invenio-mail

.. image:: https://img.shields.io/github/tag/inveniosoftware/invenio-mail.svg
        :target: https://github.com/inveniosoftware/invenio-mail/releases

.. image:: https://img.shields.io/pypi/dm/invenio-mail.svg
        :target: https://pypi.python.org/pypi/invenio-mail

.. image:: https://img.shields.io/github/license/inveniosoftware/invenio-mail.svg
        :target: https://github.com/inveniosoftware/invenio-mail/blob/master/LICENSE

Invenio-Mail is an integration layer between Invenio and Flask-Mail.

Further documentation is available on
https://invenio-mail.readthedocs.io/


CHANGES.rst
Changes
=======

Version 1.0.0b1 (released 2016-12-05)
-------------------------------------

- Initial public release.


run-tests.sh
#!/usr/bin/env sh
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

pydocstyle invenio_mail && \
# isort -rc -c -df **/*.py && \
check-manifest --ignore ".travis-*" && \
sphinx-build -qnNW docs docs/_build/html && \
python setup.py test && \
sphinx-build -qnNW -b doctest docs docs/_build/doctest


tests/conftest.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


"""Pytest configuration."""

from __future__ import absolute_import, print_function

import os
import sys
from datetime import datetime

import pytest
from flask import Blueprint, Flask
from flask_celeryext import FlaskCeleryExt


from invenio_mail import InvenioMail

PY3 = sys.version_info[0] == 3
if PY3:
    from io import StringIO
else:
    from StringIO import StringIO


@pytest.fixture(scope='session')
def email_task_app(request):
    """Flask application fixture."""
    app = Flask('testapp')
    app.config.update(
        SQLALCHEMY_DATABASE_URI=os.environ.get(
            'SQLALCHEMY_DATABASE_URI', 'sqlite://'
        ),
        CELERY_ALWAYS_EAGER=True,
        CELERY_RESULT_BACKEND='cache',
        CELERY_CACHE_BACKEND='memory',
        CELERY_EAGER_PROPAGATES_EXCEPTIONS=True,
        MAIL_SUPPRESS_SEND=True
    )
    FlaskCeleryExt(app)

    InvenioMail(app, StringIO())

    return app


@pytest.fixture(scope='session')
def email_api_app(email_task_app):
    """Flask application fixture."""
    email_task_app.register_blueprint(
        Blueprint('invenio_mail', __name__, template_folder='templates')
    )

    return email_task_app


@pytest.fixture
def email_params():
    return {
        'subject': 'subject',
        'recipients': ['recipient@inveniosoftware.com'],
        'sender': 'sender@inveniosoftware.com',
        'cc': 'cc@inveniosoftware.com',
        'bcc': 'bcc@inveniosoftware.com',
        'reply_to': 'reply_to@inveniosoftware.com',
        'date': datetime.now(),
        'attachments': [],
        'charset': None,
        'extra_headers': None,
        'mail_options': [],
        'rcpt_options': [],
    }


@pytest.fixture
def email_ctx():
    return {
        'user': 'User',
        'content': 'This a content.',
        'sender': 'sender',
    }


tests/test_invenio_mail_tasks.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


"""Module tests."""

from __future__ import absolute_import, print_function

from flask_mail import Message

from invenio_mail.tasks import send_email


def test_send_message_outbox(email_task_app):
    """Test sending a message using Task module."""
    with email_task_app.app_context():
        with email_task_app.extensions['mail'].record_messages() as outbox:
            msg = {
                'subject': 'Test1',
                'sender': 'test1@test1.test1',
                'recipients': ['test1@test1.test1']
            }

            send_email(msg)

            assert len(outbox) == 1
            sent_msg = outbox[0]
            assert sent_msg.subject == 'Test1'
            assert sent_msg.sender == 'test1@test1.test1'
            assert sent_msg.recipients == ['test1@test1.test1']


def test_send_message_stream(email_task_app):
    """Test sending a message using Task module."""
    with email_task_app.app_context():
        with email_task_app.extensions['mail'].record_messages() as outbox:
            msg = {
                'subject': 'Test2',
                'sender': 'test2@test2.test2',
                'recipients': ['test2@test2.test2']
            }

            send_email(msg)

            result_stream = email_task_app.extensions['invenio-mail'].stream
            assert result_stream.getvalue().find(
                'From: test2@test2.test2') != -1
            assert result_stream.getvalue().find('To: test2@test2.test2') != -1
            assert result_stream.getvalue().find('Subject: Test2') != -1


invenio_mail/tasks.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Background tasks for mail module."""

from __future__ import absolute_import, print_function

from celery import shared_task
from flask import current_app
from flask_celeryext import RequestContextTask
from flask_mail import Message


@shared_task(base=RequestContextTask)
def send_email(data):
    """Celery task to send emails.

    .. warning::

       Due to an incompatibility between MessagePack serialization and Message,
       support for attachments and dates is limited. Consult the tests for
       details.
    """
    msg = Message()
    msg.__dict__.update(data)
    current_app.extensions['mail'].send(msg)


invenio_mail/ext.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Invenio mail module."""

from __future__ import absolute_import, print_function

import sys
import threading

from flask_mail import Mail, email_dispatched


def print_email(message, app):
    """Print mail to stream.

    Signal handler for email_dispatched signal. Prints by default the output
    to the stream specified in the constructor of InvenioMail.

    :param message: Message object.
    :param app: Flask application object.
    """
    invenio_mail = app.extensions['invenio-mail']
    with invenio_mail._lock:
        invenio_mail.stream.write(
            '{0}\n{1}\n'.format(message.as_string(), '-' * 79))
        invenio_mail.stream.flush()


class InvenioMail(object):
    """Invenio-Mail extension."""

    def __init__(self, app=None, stream=None):
        """Extension initialization.

        Mails are only printed to the stream if ``MAIL_SUPPRESS_SEND`` is
        ``True``.

        :param app: Flask application object.
        :param stream: Stream to print emails to. Defaults to ``sys.stdout``.
        """
        self.stream = stream or sys.stdout
        self._lock = threading.RLock()
        if app:
            self.init_app(app)

    def init_app(self, app):
        """Flask application initialization.

        The initialization will:

         * Set default values for the configuration variables.
         * Initialise the Flask mail extension.
         * Configure the extension to avoid the email sending in case of debug
           or ``MAIL_SUPPRESS_SEND`` config variable set. In this case, the
           email will be written in the stream configured in the extension.

        :param app: Flask application object.
        """
        self.init_config(app)
        if 'mail' not in app.extensions:
            Mail(app)
        if app.config.get('MAIL_SUPPRESS_SEND', False) or app.debug:
            email_dispatched.connect(print_email)
        app.extensions['invenio-mail'] = self

    def init_config(self, app):
        """Initialize configuration.

        :param app: Flask application object.
        """
        app.config.setdefault('MAIL_DEBUG', app.debug)
        app.config.setdefault('MAIL_SUPPRESS_SEND', app.debug or app.testing)
